### PR TITLE
New module: elprep filter

### LINKED ISSUE #1487: new module: Elprep
I think it would be good to have a module for [Elprep](https://bioconda.github.io/recipes/elprep/README.html)
This tool offer integrated bam sort/markdup and variant calling.

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by matthdsm**
This PR add the `filter` function of elprep
Reference data for improved testing at https://github.com/nf-core/test-datasets/pull/508
## PR checklist

Closes #1487

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/elprep/filter/main.nf [hunk 1]
```diff
@@ -0,0 +1,99 @@
+process ELPREP_FILTER {
+    tag "$meta.id"
+    label 'process_high'
```
Comment (jfy133) [inline modules/elprep/filter/main.nf:3]: This seems extreme for a high performance BAM/SAM processing tool - but if in your experience it does require a lot of RAM it's fine.

File: modules/elprep/filter/main.nf [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+    // markdup args
+    if (args.contains("--mark-duplicates")){
+        args += " --mark-optical-duplicates ${prefix}.metrics.txt"
+    }
+    // variant calling args
+    if (run_haplotypecaller) {
+        args += " --haplotypecaller ${prefix}.g.vcf.gz"
+    }
+    if (reference_elfasta) {
```
Comment (FriederikeHanssen) [inline modules/elprep/filter/main.nf:57]:
```suggestion
    def fasta = reference_elfasta ? "--reference ${reference_elfasta}" : ""
```

File: modules/elprep/filter/main.nf [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    if (get_activity_regions) {
+        args += " --assembly-regions ${prefix}.assembly_regions.igv"
+    }
+    // errors
+    if ((args.contains("--bqsr") || args.contains("--haplotypecaller")) && reference_elfasta == null) error "Reference required BQSR and variant calling."
+
+    """
+    elprep filter ${bam} ${prefix}.${suffix} \\
+        --nr-of-threads $task.cpus \\
+        $args
```
Comment (FriederikeHanssen) [inline modules/elprep/filter/main.nf:90]:
```suggestion
        $fasta \\
        $known_sites \\
        $args
```

File: modules/elprep/filter/main.nf [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        args += " --mark-optical-duplicates ${prefix}.metrics.txt"
+    }
+    // variant calling args
+    if (run_haplotypecaller) {
+        args += " --haplotypecaller ${prefix}.g.vcf.gz"
+    }
+    if (reference_elfasta) {
+        args += " --reference ${reference_elfasta}"
+    }
+    if (known_sites_elsites) {
```
Comment (FriederikeHanssen) [inline modules/elprep/filter/main.nf:60]:
```suggestion
    def known_sites = known_sites_elsites ? " --known-sites ${known_sites_elsites}" : ""
```

File: modules/elprep/filter/main.nf [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        args += " --known-sites ${known_sites_elsites}"
```
Comment (FriederikeHanssen) [inline modules/elprep/filter/main.nf:61]:
```suggestion
```

File: modules/elprep/filter/main.nf [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def suffix = args.contains("--output-type sam") ? "sam" : "bam"
+
+    // filter args
+    if (reference_sequences) {
+        args += " --replace-reference-sequences ${reference_sequences}"
+    }
```
Comment (jfy133) [inline modules/elprep/filter/main.nf:46]: Any reason why all of these are not in 

```suggestion
reference_sequences ? args += " --replace-reference-sequences ${reference_sequences}" : ""
```

notation? Would make this easier to read ;)

File: modules/elprep/filter/main.nf [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    // variant calling args
+    if (run_haplotypecaller) {
+        args += " --haplotypecaller ${prefix}.g.vcf.gz"
+    }
+    if (reference_elfasta) {
+        args += " --reference ${reference_elfasta}"
+    }
+    if (known_sites_elsites) {
+        args += " --known-sites ${known_sites_elsites}"
+    }
```
Comment (FriederikeHanssen) [inline modules/elprep/filter/main.nf:62]:
```suggestion
```

File: modules/elprep/filter/main.nf [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+    // misc
+    if (get_activity_profile) {
+        args += " --activity-profile ${prefix}.activity_profile.igv"
+    }
+    if (get_activity_regions) {
+        args += " --assembly-regions ${prefix}.assembly_regions.igv"
+    }
+    // errors
+    if ((args.contains("--bqsr") || args.contains("--haplotypecaller")) && reference_elfasta == null) error "Reference required BQSR and variant calling."
```
Comment (jfy133) [inline modules/elprep/filter/main.nf:87]: Does the tool not throw this error for you?

File: modules/elprep/filter/meta.yml [hunk 9]
```diff
@@ -0,0 +1,96 @@
+name: "elprep_filter"
+description: "Filter, sort and markdup sam/bam files, with optional BQSR and variant calling."
+keywords:
+  - sort
```
Comment (jfy133) [inline modules/elprep/filter/meta.yml:4]:
```suggestion
  - sort
  - bam
  - sam
  - filter
  - variant calling
```

etc.

File: modules/elprep/filter/meta.yml [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - bam:
+      type: file
+      description: Input SAM/BAM file
+      pattern: "*.{bam,sam}"
+  - run_haplotypecaller:
+      type: boolean
+      description: Run variant callin on the input files.
```
Comment (jfy133) [inline modules/elprep/filter/meta.yml:26]:
```suggestion
      description: Run variant calling on the input files.
```

File: modules/elprep/filter/meta.yml [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+  - run_bqsr:
+      type: boolean
+      description: Run BQSR on the input files.
+  - reference_sequences:
+      type: file
+      description: Optional SAM header to replace existing header.
+  - filter_regions_bed:
```
Comment (jfy133) [inline modules/elprep/filter/meta.yml:33]: Missing pattern (as for all the other subsequent files)?

**Review by jfy133 — APPROVED** [review]
Minor formatting and documentation comments but otherwise LGTM

File: modules/elprep/filter/main.nf (hunk 1 above)
Comment (matthdsm) [reply modules/elprep/filter/main.nf:3]: Yes, It is a bit extreme, but `elprep filter` reads the entire file into ram, so it uses a lot of memory. In exchange, it's very fast. All depends on the size of the input file though.

File: modules/elprep/filter/main.nf (hunk 6 above)
Comment (matthdsm) [reply modules/elprep/filter/main.nf:46]: probably because I'm not entirely comfortable with the syntax yet. I'll modify it!

**Review by FriederikeHanssen — CHANGES_REQUESTED** [review]
I would use `def` and ternary operator mainly to comply with with coding standards in other modules similar to `fasta` and `known_Sites` for all input paths. In my opinion all the `vals` should be part of the `ext.args` and not be part of the module code to allow customization from the `modules.config` only.

File: modules/elprep/filter/main.nf (hunk 8 above)
Comment (matthdsm) [reply modules/elprep/filter/main.nf:85]: Yes it does actually, so this can be removed

**Comment by matthdsm** [comment]
Thanks for the suggestions @FriederikeHanssen. I've made some modifications as suggested by @jfy133.
The use of `vals` was also suggested by @jfy133 on [slack](https://nfcore.slack.com/archives/CJRH30T6V/p1649851420932219?thread_ts=1649850693.962269&cid=CJRH30T6V), so I don't know what's the best route to go.

File: modules/elprep/filter/meta.yml [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - bam:
+      type: file
+      description: Input SAM/BAM file
+      pattern: "*.{bam,sam}"
+  - run_haplotypecaller:
+      type: boolean
+      description: Run variant calling on the input files.
```
Comment (FriederikeHanssen) [inline modules/elprep/filter/meta.yml:30]: Maybe add here: "Needed to generate output `gvcf` etc.

**Review by FriederikeHanssen — APPROVED** [review]
After the discussion on slack, other then maybe an extra line to the meta yml about the optional outputs and their relation to the input vals, I am happy with this.

File: modules/elprep/filter/main.nf [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+    // bqsr args
+    run_bqsr ? args += " --bqsr ${prefix}.recall": ""
+
+    intermediate_bqsr_tables ? args += " --bqsr-apply .": ""
+    bqsr_tables_only ? args += " --bqsr-tables-only ${prefix}.table": ""
+
+    // misc
+    get_activity_profile ? args += " --activity-profile ${prefix}.activity_profile.igv": ""
+    get_assembly_regions ? args += " --assembly-regions ${prefix}.assembly_regions.igv": ""
```
Comment (FriederikeHanssen) [inline modules/elprep/filter/main.nf:64]: this is a bit unusual to see, but i am not to fussed about it. Normally args just contains whatever is in `ext.args` and all the extra `def` things are added explicetly

File: modules/elprep/filter/main.nf (hunk 2 above)
Comment (matthdsm) [reply modules/elprep/filter/main.nf]: fixed this everywhere!

File: modules/elprep/filter/main.nf (hunk 13 above)
Comment (matthdsm) [reply modules/elprep/filter/main.nf]: `def`ed all the things

**Review by jfy133 — APPROVED** [review]
Good to go!

**Comment by matthdsm** [comment]
Thanks @jfy133 and @FriederikeHanssen !

**Pull request merged**